builder/tart/vnc.go

```diff
@@ -0,0 +1,285 @@
+		d.waitString.Grow(1)
+		return nil
+	case WaitForStringEnd:
+		waitString := d.waitString.String()
+		d.waitString.Reset()
+
+		waitStringCStr := C.CString(waitString)
+		defer C.free(unsafe.Pointer(waitStringCStr))
+
+		for {
+			fmt.Fprintf(os.Stderr, "Waiting for '%s'...\n", waitString)
+			if err := d.WaitForFramebufferUpdate(); err != nil {
+				return err
+			}
+
+			if C.recognizeTextInFramebuffer(waitStringCStr,
+				unsafe.Pointer(&d.frameBuffer.Pix[0]),
+				C.int(d.frameBuffer.Bounds().Dx()),
+				C.int(d.frameBuffer.Bounds().Dy())) {
+				break
+			}
+		}
+
+		return nil
+	default:
+		if d.waitString.Cap() > 0 {
+			d.waitString.WriteRune(key)
+			return nil
+		} else {
+			return d.vncDriver.SendKey(key, action)
+		}
+	}
+}
+
+func (d *customDriver) SendSpecial(special string, action bootcommand.KeyAction) error {
+	return d.vncDriver.SendSpecial(special, action)
+}
+
+func (d *customDriver) Flush() error {
+	return d.vncDriver.Flush()
+}
+
+func (d *customDriver) WaitForFramebufferUpdate() error {
+
+	incremental := true
+	for {
+		w, h := d.vncClient.FrameBufferWidth, d.vncClient.FrameBufferHeight
+		fmt.Fprintf(os.Stderr, "Requesting frame buffer update for %dx%d\n", w, h)
+
+		if err := d.vncClient.FramebufferUpdateRequest(incremental, 0, 0, w, h); err != nil {
+			return err
+		}
+
+		select {
+		case msg := <-d.serverMessageChannel:
+			fmt.Fprintln(os.Stderr, "Received message type", msg.Type(), msg)
+			if msg.Type() == 0 {
+				framebufferUpdateMessage := msg.(*vnc.FramebufferUpdateMessage)
+
+				if len(framebufferUpdateMessage.Rectangles) == 0 {
+					return fmt.Errorf("Frame update did not have any rectangles")
+				}
+
+				for _, rect := range framebufferUpdateMessage.Rectangles {
+					switch encoding := rect.Enc.(type) {
+					case *DesktopSizePseudoEncoding:
+						w, h := int(d.vncClient.FrameBufferWidth), int(d.vncClient.FrameBufferHeight)
+						d.frameBuffer = image.NewRGBA(image.Rect(0, 0, w, h))
+						fmt.Fprintf(os.Stderr, "New desktop size is %dx%d, resized framebuffer\n", w, h)
+						continue
+					case *ZlibEncoding:
+						for i, c := range encoding.Colors {
+							x, y := i%int(rect.Width), i/int(rect.Width)
+							r, g, b := uint8(c.R), uint8(c.G), uint8(c.B)
+							d.frameBuffer.Set(int(rect.X)+x, int(rect.Y)+y, color.RGBA{r, g, b, 255})
+						}
+					default:
+						return fmt.Errorf("Frame had unknown encoding %s", encoding)
+					}
+				}
+				return nil
+			} else {
+				// Ignore messages we didn't ask for
+				continue
+			}
+		case <-d.ctx.Done():
+			return d.ctx.Err()
+		}
+	}
+}
+
+type DesktopSizePseudoEncoding struct{}
+
+func (*DesktopSizePseudoEncoding) Read(c *vnc.ClientConn, rect *vnc.Rectangle, r io.Reader) (vnc.Encoding, error) {
+	c.FrameBufferWidth = rect.Width
+	c.FrameBufferHeight = rect.Height
+	return &DesktopSizePseudoEncoding{}, nil
+}
+
+func (*DesktopSizePseudoEncoding) Type() int32 {
+	return -223 // RFC 6143 7.8.2
+}
+
+type ZlibEncoding struct {
```

Comment:
Why use and implement `zlib` encoding when the `Raw` encoding [is already implemented](https://pkg.go.dev/github.com/alexsnet/go-vnc#RawEncoding) in the VNC library?

It feels that when communicating with the VMs on a local machine, code complexity factor is more important the the bandwidth implications.

Reply:
In my testing the `Raw` encoding was performing significantly slower in picking up new frames from the VM than with `zlib`. I'll do some more testing. If the performance is the same we don't need `zlib`.